pkg/ebpf/c/tracee.bpf.c

```diff
@@ -970,12 +1033,29 @@ statfunc int walk_mod_tree(program_data_t *p, struct rb_node *root, int idx)
         }
     }
 
-    return HID_MOD_UNCOMPLETED_ITERATIONS;
-}
+    iteration_num++;
 
-struct mod_tree_root {
-    struct latch_tree_root root;
-};
+    if (loop_result == HID_MOD_COMPLETED_ITERATIONS) {
+        flags = HISTORY_SCAN_FINISHED;
+        lkm_seeker_send_to_userspace((struct module *) HISTORY_SCAN_SUCCESSFUL, &flags, &p);
+        bpf_tail_call(ctx, &prog_array, TAIL_HIDDEN_KERNEL_MODULE_PROC);
+    } else if (loop_result == HID_MOD_RACE_CONDITION || iteration_num == MOD_TREE_LOOP_DEPTH) {
+        flags = HISTORY_SCAN_FINISHED;
+        tracee_log(ctx, BPF_LOG_LVL_WARN, BPF_LOG_ID_HID_KER_MOD, loop_result ^ iteration_num);
+        lkm_seeker_send_to_userspace((struct module *) HISTORY_SCAN_FAILURE, &flags, &p);
+        bpf_tail_call(ctx, &prog_array, TAIL_HIDDEN_KERNEL_MODULE_PROC);
```

Comment:
This line can be extracted out of these if-else statements

Reply:
In case the if else statements are not fulfilled, it updates the map and does another recursive tail call.
If I extract this tail call (Line 1046) out of the if-else statement, there won't be the map update and further recursive call).

Reply:
What do you mean by "not fulfilled"?
I was refering to the fact that both the if block and else block contains this same tail call at the end of the block, so I don't see a reason why not to take it out:
```
   if (loop_result == HID_MOD_COMPLETED_ITERATIONS) {
        flags = HISTORY_SCAN_FINISHED;
        lkm_seeker_send_to_userspace((struct module *) HISTORY_SCAN_SUCCESSFUL, &flags, &p);
    } else if (loop_result == HID_MOD_RACE_CONDITION || iteration_num == MOD_TREE_LOOP_DEPTH) {
        flags = HISTORY_SCAN_FINISHED;
        tracee_log(ctx, BPF_LOG_LVL_WARN, BPF_LOG_ID_HID_KER_MOD, loop_result ^ iteration_num);
        lkm_seeker_send_to_userspace((struct module *) HISTORY_SCAN_FAILURE, &flags, &p);
    }
bpf_tail_call(ctx, &prog_array, TAIL_HIDDEN_KERNEL_MODULE_PROC);
```

Reply:
As said, in that case the code highlighted below won't happen (note that it's not if-else.. it's if and else if):
![image](https://github.com/user-attachments/assets/b3ad419f-1c7d-4d15-a020-d639f44dc962)